File: dsa-python/solutions/sliding_window/medium/permutation_in_string.py
# 567. Permutation in String
# https://leetcode.com/problems/permutation-in-string/
#
# Given two strings s1 and s2, return true if s2 contains a permutation of s1,
# or false otherwise. In other words, return true if one of s1's permutations
# is the substring of s2.
#
# APPROACH: Fixed sliding window of size len(s1) with a "matches" counter.
# Build frequency arrays for s1 and the first window of s2, count how many of
# 26 characters have equal frequency. Slide window one char at a time, updating
# matches in O(1) per step. When matches == 26, a permutation is found.
#
# TIME: O(n) where n = len(s2)
# SPACE: O(1) — two arrays of size 26


class PermutationInString:
    def check_inclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2):
            return False

        s1_count = [0] * 26
        s2_count = [0] * 26

        for i in range(len(s1)):
            s1_count[ord(s1[i]) - ord('a')] += 1
            s2_count[ord(s2[i]) - ord('a')] += 1

        # count how many of 26 characters already match
        matches = 0
        for i in range(26):
            if s1_count[i] == s2_count[i]:
                matches += 1

        # slide the window
        l = 0
        for r in range(len(s1), len(s2)):
            if matches == 26:
                return True

            # add right character
            idx = ord(s2[r]) - ord('a')
            s2_count[idx] += 1
            if s2_count[idx] == s1_count[idx]:
                matches += 1
            elif s2_count[idx] == s1_count[idx] + 1:
                matches -= 1

            # remove left character
            idx = ord(s2[l]) - ord('a')
            s2_count[idx] -= 1
            if s2_count[idx] == s1_count[idx]:
                matches += 1
            elif s2_count[idx] == s1_count[idx] - 1:
                matches -= 1

            l += 1

        return matches == 26

File: dsa-csharp/solutions/SlidingWindow/Medium/PermutationInString.cs
// 567. Permutation in String
// https://leetcode.com/problems/permutation-in-string/
//
// Given two strings s1 and s2, return true if s2 contains a permutation of s1,
// or false otherwise. In other words, return true if one of s1's permutations
// is the substring of s2.
//
// APPROACH: Fixed sliding window of size len(s1) with a "matches" counter.
// Build frequency arrays for s1 and the first window of s2, count how many of
// 26 characters have equal frequency. Slide window one char at a time, updating
// matches in O(1) per step. When matches == 26, a permutation is found.
//
// TIME: O(n) where n = len(s2)
// SPACE: O(1) — two arrays of size 26

public class PermutationInString
{
    public bool CheckInclusion(string s1, string s2)
    {
        if(s1.Length > s2.Length) return false;

        var s1_count = new int [26];
        var s2_count = new int [26];

        for(int i = 0 ; i<s1.Length; i++)
        {
            s1_count[s1[i] - 'a']++;
            s2_count[s2[i] - 'a']++;
        }

        int matches = 0;

        for(int i =0; i < s1_count.Length; i++)
        {
            if(s1_count[i] == s2_count[i]) matches++;
        }

        int left = 0;

        for(int r = s1.Length; r < s2.Length; r++)
        {
            if(matches == 26) return true;

            //Add a right Character
            int index = s2[r] - 'a'; //index of array for the new right character
            s2_count[index]++; //updated the frequency of new right character

            if(s2_count[index] == s1_count[index]) matches++;

            else if(s2_count[index] == s1_count[index] + 1) matches--; //If it does not equate now means matches dropped by one

            //Same thing for left when we remove a character

            index = s2[left] - 'a';
            s2_count[index]--; //Removing the character from window, Reducing the frequency by one

            if(s2_count[index] == s1_count[index]) matches++;

            else if(s2_count[index] == s1_count[index] - 1) matches--; //Means removing the character has made the frequencies unequal, that means it will not match

            left++;

        }

        return matches == 26;


    }
}
